## Human Activity Recognition 

<br>


This project is to build a model that predicts the human activities such as Walking, Walking_Upstairs, Walking_Downstairs, Sitting, Standing or Laying.

This dataset is collected from 30 persons(referred as subjects in this dataset), performing different activities with a smartphone to their waists. The data is recorded with the help of sensors (accelerometer and Gyroscope) in that smartphone. This experiment was video recorded to label the data manually.

## How data was recorded

By using the sensors(Gyroscope and accelerometer) in a smartphone, they have captured '3-axial linear acceleration'(_tAcc-XYZ_) from accelerometer and '3-axial angular velocity' (_tGyro-XYZ_) from Gyroscope with several variations. 

> prefix 't' in those metrics denotes time.

> suffix 'XYZ' represents 3-axial signals in X , Y, and Z directions.

### Feature names

1. These sensor signals are preprocessed by applying noise filters and then sampled in fixed-width windows(sliding windows) of 2.56 seconds each with 50% overlap. ie., each window has 128 readings. 

2. From Each window, a feature vector was obtianed by calculating variables from the time and frequency domain.
> In our dataset, each datapoint represents a window with different readings 
3. The accelertion signal was saperated into Body and Gravity acceleration signals(___tBodyAcc-XYZ___ and ___tGravityAcc-XYZ___) using some low pass filter with corner frequecy of 0.3Hz.

4. After that, the body linear acceleration and angular velocity were derived in time to obtian _jerk signals_ (___tBodyAccJerk-XYZ___ and ___tBodyGyroJerk-XYZ___). 

5. The magnitude of these 3-dimensional signals were calculated using the Euclidian norm. This magnitudes are represented as features with names like _tBodyAccMag_, _tGravityAccMag_, _tBodyAccJerkMag_, _tBodyGyroMag_ and _tBodyGyroJerkMag_.

6. Finally, We've got frequency domain signals from some of the available signals by applying a FFT (Fast Fourier Transform). These signals obtained were labeled with ___prefix 'f'___ just like original signals with ___prefix 't'___. These signals are labeled as ___fBodyAcc-XYZ___, ___fBodyGyroMag___ etc.,.

7. These are the signals that we got so far.
	+ tBodyAcc-XYZ
	+ tGravityAcc-XYZ
	+ tBodyAccJerk-XYZ
	+ tBodyGyro-XYZ
	+ tBodyGyroJerk-XYZ
	+ tBodyAccMag
	+ tGravityAccMag
	+ tBodyAccJerkMag
	+ tBodyGyroMag
	+ tBodyGyroJerkMag
	+ fBodyAcc-XYZ
	+ fBodyAccJerk-XYZ
	+ fBodyGyro-XYZ
	+ fBodyAccMag
	+ fBodyAccJerkMag
	+ fBodyGyroMag
	+ fBodyGyroJerkMag

8. We can esitmate some set of variables from the above signals. ie., We will estimate the following properties on each and every signal that we recoreded so far.

	+ ___mean()___: Mean value
	+ ___std()___: Standard deviation
	+ ___mad()___: Median absolute deviation 
	+ ___max()___: Largest value in array
	+ ___min()___: Smallest value in array
	+ ___sma()___: Signal magnitude area
	+ ___energy()___: Energy measure. Sum of the squares divided by the number of values. 
	+ ___iqr()___: Interquartile range 
	+ ___entropy()___: Signal entropy
	+ ___arCoeff()___: Autorregresion coefficients with Burg order equal to 4
	+ ___correlation()___: correlation coefficient between two signals
	+ ___maxInds()___: index of the frequency component with largest magnitude
	+ ___meanFreq()___: Weighted average of the frequency components to obtain a mean frequency
	+ ___skewness()___: skewness of the frequency domain signal 
	+ ___kurtosis()___: kurtosis of the frequency domain signal 
	+ ___bandsEnergy()___: Energy of a frequency interval within the 64 bins of the FFT of each window.
	+ ___angle()___: Angle between to vectors.

9. We can obtain some other vectors by taking the average of signals in a single window sample. These are used on the angle() variable'
`
	+ gravityMean
	+ tBodyAccMean
	+ tBodyAccJerkMean
	+ tBodyGyroMean
	+ tBodyGyroJerkMean


###  Y_Labels(Encoded)
+ In the dataset, Y_labels are represented as numbers from 1 to 6 as their identifiers.

	- WALKING as __1__
	- WALKING_UPSTAIRS as __2__
	- WALKING_DOWNSTAIRS as __3__
	- SITTING as __4__
	- STANDING as __5__
	- LAYING as __6__
    
## Train and test data were saperated
 - The readings from ___70%___ of the volunteers were taken as ___trianing data___ and remaining ___30%___ subjects recordings were taken for ___test data___
 
## Data

* All the data is present in 'UCI_HAR_dataset/' folder in present working directory.
     - Feature names are present in 'UCI_HAR_dataset/features.txt'
     - ___Train Data___
         - 'UCI_HAR_dataset/train/X_train.txt'
         - 'UCI_HAR_dataset/train/subject_train.txt'
         - 'UCI_HAR_dataset/train/y_train.txt'
     - ___Test Data___
         - 'UCI_HAR_dataset/test/X_test.txt'
         - 'UCI_HAR_dataset/test/subject_test.txt'
         - 'UCI_HAR_dataset/test/y_test.txt'
         

## Data Size :
> 27 MB


# Quick overview of the dataset :



* Accelerometer and Gyroscope readings are taken from 30 volunteers(referred as subjects) while performing the following 6 Activities.

    1. Walking     
    2. WalkingUpstairs 
    3. WalkingDownstairs 
    4. Standing 
    5. Sitting 
    6. Lying.


* Readings are divided into a window of 2.56 seconds with 50% overlapping. 

* Accelerometer readings are divided into gravity acceleration and body acceleration readings,
  which has x,y and z components each.

* Gyroscope readings are the measure of angular velocities which has x,y and z components.

* Jerk signals are calculated for BodyAcceleration readings.

* Fourier Transforms are made on the above time readings to obtain frequency readings.

* Now, on all the base signal readings., mean, max, mad, sma, arcoefficient, engerybands,entropy etc., are calculated for each window.

* We get a feature vector of 561 features and these features are given in the dataset.

* Each window of readings is a datapoint of 561 features.

## Problem Framework

* 30 subjects(volunteers) data is randomly split to 70%(21) test and 30%(7) train data.
* Each datapoint corresponds one of the 6 Activities.


## Problem Statement

 + Given a new datapoint we have to predict the Activity

In [1]:
import pandas as pd
import numpy as np

In [2]:
# Activities are the class labels
# It is a 6 class classification
ACTIVITIES = {
    0: 'WALKING',
    1: 'WALKING_UPSTAIRS',
    2: 'WALKING_DOWNSTAIRS',
    3: 'SITTING',
    4: 'STANDING',
    5: 'LAYING',
}

# Utility function to print the confusion matrix
def confusion_matrix(Y_true, Y_pred):
    Y_true = pd.Series([ACTIVITIES[y] for y in np.argmax(Y_true, axis=1)])
    Y_pred = pd.Series([ACTIVITIES[y] for y in np.argmax(Y_pred, axis=1)])

    return pd.crosstab(Y_true, Y_pred, rownames=['True'], colnames=['Pred'])

### Data

In [3]:
# Data directory
DATADIR = 'UCI_HAR_Dataset'

In [4]:
# Raw data signals
# Signals are from Accelerometer and Gyroscope
# The signals are in x,y,z directions
# Sensor signals are filtered to have only body acceleration
# excluding the acceleration due to gravity
# Triaxial acceleration from the accelerometer is total acceleration
SIGNALS = [
    "body_acc_x",
    "body_acc_y",
    "body_acc_z",
    "body_gyro_x",
    "body_gyro_y",
    "body_gyro_z",
    "total_acc_x",
    "total_acc_y",
    "total_acc_z"
]

In [5]:
# Utility function to read the data from csv file
def _read_csv(filename):
    return pd.read_csv(filename, delim_whitespace=True, header=None)

# Utility function to load the load
def load_signals(subset):
    signals_data = []

    for signal in SIGNALS:
        filename = f'UCI_HAR_Dataset/{subset}/Inertial Signals/{signal}_{subset}.txt'
        signals_data.append(
            _read_csv(filename).as_matrix()
        ) 

    # Transpose is used to change the dimensionality of the output,
    # aggregating the signals by combination of sample/timestep.
    # Resultant shape is (7352 train/2947 test samples, 128 timesteps, 9 signals)
    return np.transpose(signals_data, (1, 2, 0))

In [6]:

def load_y(subset):
    """
    The objective that we are trying to predict is a integer, from 1 to 6,
    that represents a human activity. We return a binary representation of 
    every sample objective as a 6 bits vector using One Hot Encoding
    (https://pandas.pydata.org/pandas-docs/stable/generated/pandas.get_dummies.html)
    """
    filename = f'UCI_HAR_Dataset/{subset}/y_{subset}.txt'
    y = _read_csv(filename)[0]

    return pd.get_dummies(y).as_matrix()

In [7]:
def load_data():
    """
    Obtain the dataset from multiple files.
    Returns: X_train, X_test, y_train, y_test
    """
    X_train, X_test = load_signals('train'), load_signals('test')
    y_train, y_test = load_y('train'), load_y('test')

    return X_train, X_test, y_train, y_test

In [8]:
# Importing tensorflow
np.random.seed(42)
import tensorflow as tf
tf.logging.set_verbosity(tf.logging.ERROR)
tf.set_random_seed(42)

In [9]:
# Configuring a session
session_conf = tf.ConfigProto(
    intra_op_parallelism_threads=1,
    inter_op_parallelism_threads=1
)

In [10]:
# Import Keras
from keras import backend as K
sess = tf.Session(graph=tf.get_default_graph(), config=session_conf)
K.set_session(sess)

Using TensorFlow backend.


In [11]:
# Importing libraries
import keras
from keras.models import Sequential
from keras.layers import LSTM, BatchNormalization, Flatten
from keras.layers.core import Dense, Dropout

In [12]:
# Loading the train and test data
X_train, X_test, Y_train, Y_test = load_data()

C:\Users\vivek\Anaconda3\lib\site-packages\ipykernel_launcher.py:12: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.
  if sys.path[0] == '':


In [13]:
# Utility function to count the number of classes
def _count_classes(y):
    return len(set([tuple(category) for category in y]))

In [14]:
timesteps = len(X_train[0])
input_dim = len(X_train[0][0])
n_classes = _count_classes(Y_train)

print(timesteps)
print(input_dim)
print(len(X_train))

128
9
7352


- Defining the Architecture of LSTM

In [15]:
import warnings
warnings.filterwarnings('ignore')
X_train  =  load_signals('train')
Y_train  =  load_y('train')
X_test  =  load_signals('test')
Y_test  =  load_y('test')

In [16]:
y_int_train = np.array([np.where(i == 1)[0][0]+1  for i in Y_train])
y_int_test = np.array([np.where(i == 1)[0][0]+1  for i in Y_test])

In [17]:
y_train =  pd.DataFrame(y_int_train, columns = ['label'])
x_train = pd.DataFrame(X_train.reshape(X_train.shape[0], 128*9))

y_test =  pd.DataFrame(y_int_test, columns = ['label'])
x_test = pd.DataFrame(X_test.reshape(X_test.shape[0] , 128*9))

data_train = pd.DataFrame(pd.concat([x_train,y_train], axis =1))
data_test = pd.DataFrame(pd.concat([x_test,y_test], axis =1))

In [18]:
data_dynamic_train = data_train[data_train['label'].isin([1,2,3])]
data_dynamic_test = data_test[data_test['label'].isin([1,2,3])]

data_static_train = data_train[data_train['label'].isin([4,5,6])]
data_static_test = data_test[data_test['label'].isin([4,5,6])]

In [19]:
import warnings
warnings.filterwarnings('ignore')
data_dynamic_train['label'] = 1
data_dynamic_test['label'] = 1

data_static_train['label'] = 0
data_static_test['label'] = 0

In [20]:
from sklearn.utils import shuffle
data_dynsta_train = shuffle(pd.concat([data_dynamic_train, data_static_train],axis =0))
data_dynsta_test = shuffle(pd.concat([data_dynamic_test, data_static_test],axis =0))

In [21]:
X_dynsta_train = np.array(data_dynsta_train.drop(['label'],axis=1)).reshape(data_dynsta_train.shape[0],128,9)
Y_dynsta_train = data_dynsta_train['label']

X_dynsta_test = np.array(data_dynsta_test.drop(['label'],axis=1)).reshape(data_dynsta_test.shape[0],128,9)
Y_dynsta_test = data_dynsta_test['label']

In [22]:
from keras.utils import to_categorical
Y_dynsta_train = to_categorical(Y_dynsta_train)
Y_dynsta_test = to_categorical(Y_dynsta_test)

In [23]:
from keras.layers import Conv1D,Add,Input, Activation,MaxPooling1D, Concatenate, MaxPooling2D ,Reshape, Conv2D, Flatten, Bidirectional

In [24]:
model1 = Sequential()

model1.add(Conv1D(64, input_shape = (128, 9), kernel_size = 4,  strides = 2, kernel_initializer = 'lecun_normal', padding = 'same'))
model1.add(Dropout(0.7))

model1.add(LSTM(100,  activation = 'softsign',recurrent_dropout = 0.2, kernel_initializer = 'lecun_normal',  dropout=0.5))
model1.add(Dropout(0.85))

model1.add(Dense(2, activation = 'softmax'))
model1.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv1d_1 (Conv1D)            (None, 64, 64)            2368      
_________________________________________________________________
dropout_1 (Dropout)          (None, 64, 64)            0         
_________________________________________________________________
lstm_1 (LSTM)                (None, 100)               66000     
_________________________________________________________________
dropout_2 (Dropout)          (None, 100)               0         
_________________________________________________________________
dense_1 (Dense)              (None, 2)                 202       
Total params: 68,570
Trainable params: 68,570
Non-trainable params: 0
_________________________________________________________________


In [25]:
# Compiling the model
model1.compile(loss='categorical_crossentropy',
              optimizer=keras.optimizers.adam(lr = 0.01),
              metrics=['accuracy'])

In [26]:
from keras.callbacks import ReduceLROnPlateau, EarlyStopping

earlystop = EarlyStopping(monitor = 'val_loss', mode = 'min', patience = 20)
reducelr = ReduceLROnPlateau(monitor = 'val_loss',rate = 0.95,mode='min',verbose = 1,patience = 15)

In [27]:
# Training the model
model1.fit(X_dynsta_train,
          Y_dynsta_train,
          batch_size=4096,
          validation_data=(X_dynsta_test, Y_dynsta_test),
          epochs=50,
          verbose=1)

Train on 7352 samples, validate on 2947 samples
Epoch 1/50
7352/7352 [==============================] - 3s 434us/step - loss: 0.6973 - acc: 0.5666 - val_loss: 0.5142 - val_acc: 0.7004
Epoch 2/50
7352/7352 [==============================] - 1s 107us/step - loss: 0.5386 - acc: 0.7013 - val_loss: 0.5412 - val_acc: 0.7730
Epoch 3/50
7352/7352 [==============================] - 1s 108us/step - loss: 0.4286 - acc: 0.7850 - val_loss: 1.0835 - val_acc: 0.8018
Epoch 4/50
7352/7352 [==============================] - 1s 107us/step - loss: 0.3546 - acc: 0.8543 - val_loss: 0.2787 - val_acc: 0.8890
Epoch 5/50
7352/7352 [==============================] - 1s 108us/step - loss: 0.2520 - acc: 0.9158 - val_loss: 0.5794 - val_acc: 0.7852
Epoch 6/50
7352/7352 [==============================] - 1s 107us/step - loss: 0.1651 - acc: 0.9550 - val_loss: 0.1797 - val_acc: 0.9498
Epoch 7/50
7352/7352 [==============================] - 1s 107us/step - loss: 0.1089 - acc: 0.9735 - val_loss: 0.1267 - val_acc: 0.9640


In [28]:
model1.save('model1.h5')

In [24]:
from keras.models import load_model

model1 = load_model('model1.h5')

## Submodel(Dynamic):

In [24]:
data_dynamic_train = shuffle(data_train[data_train['label'].isin([1,2,3])])
data_dynamic_test = shuffle(data_test[data_test['label'].isin([1,2,3])])

data_static_train = shuffle(data_train[data_train['label'].isin([4,5,6])])
data_static_test = shuffle(data_test[data_test['label'].isin([4,5,6])])

In [25]:
data_static_train['label'] = data_static_train['label'].replace([4,5,6],[0,1,2])
data_static_test['label'] = data_static_test['label'].replace([4,5,6],[0,1,2])
data_dynamic_train['label'] = data_dynamic_train['label'].replace([1,2,3],[0,1,2])
data_dynamic_test['label'] = data_dynamic_test['label'].replace([1,2,3],[0,1,2])


In [26]:
x_dynamic_train = np.array(data_dynamic_train.drop(['label'], axis = 1)).reshape(data_dynamic_train.shape[0],128,9)
x_dynamic_test = np.array(data_dynamic_test.drop(['label'], axis = 1)).reshape(data_dynamic_test.shape[0],128,9)
x_static_train = np.array(data_static_train.drop(['label'], axis = 1)).reshape(data_static_train.shape[0],128,9)
x_static_test = np.array(data_static_test.drop(['label'], axis = 1)).reshape(data_static_test.shape[0],128,9)

In [27]:
y_dynamic_train = to_categorical(data_dynamic_train['label'])
y_dynamic_test = to_categorical(data_dynamic_test['label'])
y_static_train = to_categorical(data_static_train['label'])
y_static_test = to_categorical(data_static_test['label'])

In [28]:
model2 = Sequential()

model2.add(Conv1D(48, input_shape = (128, 9), kernel_size = 4,  strides = 2, kernel_initializer = 'lecun_normal', padding = 'same'))
model2.add(Dropout(0.65))

model2.add(LSTM(100,  activation = 'softsign',recurrent_dropout = 0.2, kernel_initializer = 'lecun_normal',  dropout=0.5))
model2.add(Dropout(0.8))

model2.add(Dense(3, activation = 'softmax'))
model2.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv1d_1 (Conv1D)            (None, 64, 48)            1776      
_________________________________________________________________
dropout_1 (Dropout)          (None, 64, 48)            0         
_________________________________________________________________
lstm_1 (LSTM)                (None, 100)               59600     
_________________________________________________________________
dropout_2 (Dropout)          (None, 100)               0         
_________________________________________________________________
dense_1 (Dense)              (None, 3)                 303       
Total params: 61,679
Trainable params: 61,679
Non-trainable params: 0
_________________________________________________________________


In [243]:
# Compiling the model
model2.compile(loss='categorical_crossentropy',
              optimizer=keras.optimizers.adam(lr = 0.01),
              metrics=['accuracy'])

In [244]:
# Training the model
model2.fit(x_dynamic_train,
          y_dynamic_train,
          batch_size=4096,
          validation_data=(x_dynamic_test, y_dynamic_test),
          epochs=100,
          verbose=1)

Train on 3285 samples, validate on 1387 samples
Epoch 1/100
3285/3285 [==============================] - 52s 16ms/step - loss: 1.1821 - acc: 0.3230 - val_loss: 1.1186 - val_acc: 0.3576
Epoch 2/100
3285/3285 [==============================] - 0s 128us/step - loss: 1.1348 - acc: 0.3601 - val_loss: 1.1125 - val_acc: 0.3576
Epoch 3/100
3285/3285 [==============================] - 0s 120us/step - loss: 1.1133 - acc: 0.3775 - val_loss: 1.0787 - val_acc: 0.3576
Epoch 4/100
3285/3285 [==============================] - 0s 127us/step - loss: 1.0972 - acc: 0.3921 - val_loss: 1.0490 - val_acc: 0.3886
Epoch 5/100
3285/3285 [==============================] - 0s 123us/step - loss: 1.0802 - acc: 0.4079 - val_loss: 1.0320 - val_acc: 0.4737
Epoch 6/100
3285/3285 [==============================] - 0s 117us/step - loss: 1.0730 - acc: 0.4152 - val_loss: 1.0254 - val_acc: 0.5321
Epoch 7/100
3285/3285 [==============================] - 0s 126us/step - loss: 1.0579 - acc: 0.4332 - val_loss: 1.0196 - val_acc: 

3285/3285 [==============================] - 0s 123us/step - loss: 0.0839 - acc: 0.9781 - val_loss: 0.1135 - val_acc: 0.9625
Epoch 61/100
3285/3285 [==============================] - 0s 121us/step - loss: 0.0948 - acc: 0.9732 - val_loss: 0.1150 - val_acc: 0.9618
Epoch 62/100
3285/3285 [==============================] - 0s 125us/step - loss: 0.0801 - acc: 0.9753 - val_loss: 0.1333 - val_acc: 0.9603
Epoch 63/100
3285/3285 [==============================] - 0s 121us/step - loss: 0.0847 - acc: 0.9738 - val_loss: 0.1324 - val_acc: 0.9603
Epoch 64/100
3285/3285 [==============================] - 0s 123us/step - loss: 0.0778 - acc: 0.9760 - val_loss: 0.1275 - val_acc: 0.9589
Epoch 65/100
3285/3285 [==============================] - 0s 118us/step - loss: 0.0848 - acc: 0.9747 - val_loss: 0.1161 - val_acc: 0.9632
Epoch 66/100
3285/3285 [==============================] - 0s 121us/step - loss: 0.0694 - acc: 0.9784 - val_loss: 0.0912 - val_acc: 0.9690
Epoch 67/100
3285/3285 [=======================

In [251]:
model2.save('model2.h5')

In [25]:
from keras.models import load_model

model2 = load_model('model2.h5')

## Submodel( Static )

In [47]:
from keras.layers import MaxPooling1D

In [228]:
model3 = Sequential()

model3.add(Conv1D(48, input_shape = (128, 9), kernel_size = 11,  strides = 2, kernel_initializer = 'lecun_normal', padding = 'same'))
model3.add(MaxPooling1D(5))
model3.add(Conv1D(64, kernel_size = 2,  strides = 1, kernel_initializer = 'lecun_normal', activation='relu'))
model3.add(MaxPooling1D(2))
model3.add(Conv1D(64, kernel_size = 2,  strides = 1, kernel_initializer = 'lecun_normal', activation='relu'))
model3.add(MaxPooling1D(2))
model3.add(Dropout(0.975))

model3.add(Flatten())
model3.add(Dense(3, activation = 'softmax'))
model3.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv1d_166 (Conv1D)          (None, 64, 48)            4800      
_________________________________________________________________
max_pooling1d_142 (MaxPoolin (None, 12, 48)            0         
_________________________________________________________________
conv1d_167 (Conv1D)          (None, 11, 64)            6208      
_________________________________________________________________
max_pooling1d_143 (MaxPoolin (None, 5, 64)             0         
_________________________________________________________________
conv1d_168 (Conv1D)          (None, 4, 64)             8256      
_________________________________________________________________
max_pooling1d_144 (MaxPoolin (None, 2, 64)             0         
_________________________________________________________________
dropout_114 (Dropout)        (None, 2, 64)             0         
__________

In [229]:
# Compiling the model
model3.compile(loss='categorical_crossentropy',
              optimizer=keras.optimizers.adam(lr = 0.01),
              metrics=['accuracy'])

In [233]:
# Training the model
model3.fit(x_static_train,
          y_static_train,
          batch_size=4096,
          validation_data=(x_static_test, y_static_test),
          epochs=2000,
          verbose=0)

In [234]:
model3.evaluate(x_static_train, y_static_train, batch_size =4096)

4067/4067 [==============================] - 0s 13us/step


[0.09474246203899384, 0.9798377156257629]

In [235]:
model3.evaluate(x_static_test, y_static_test, batch_size =4096)

1560/1560 [==============================] - 0s 13us/step


[0.14470736682415009, 0.9557692408561707]

In [236]:
model3.save('model3.h5')

In [26]:
from keras.models import load_model

model3 = load_model('model3.h5')

### Final prediction using above 3 models

In [53]:
import numpy as np
def predict(x):
    
    predictions = []
    
    dynOrStat = np.argmax(model1.predict(x), axis = 1)
    
    x_dyn = x[dynOrStat == 1]
    x_sta = x[dynOrStat == 0]
    
    predict_dynamic = np.argmax(model2.predict(x_dyn), axis = 1)
    predict_static = np.argmax(model3.predict(x_sta), axis = 1)+3 
    
    i = 0
    j = 0
    for output in dynOrStat:
        if output == 1:
            predictions.append(predict_dynamic[i])
            i = i + 1
        else:
            predictions.append(predict_static[j])
            j = j + 1
    
    return predictions

In [54]:
def confusion_matrix(Y_true, Y_pred):
    Y_true = pd.Series([ACTIVITIES[y] for y in Y_true])
    Y_pred = pd.Series([ACTIVITIES[y] for y in Y_pred])

    return pd.crosstab(Y_true, Y_pred, rownames=['True'], colnames=['Pred'])

In [55]:
print(confusion_matrix(np.argmax(Y_test, axis =1 ), predict(X_test)))

Pred                LAYING  SITTING  STANDING  WALKING  WALKING_DOWNSTAIRS  \
True                                                                         
LAYING                 536        0         0        1                   0   
SITTING                  3      441        46        0                   0   
STANDING                 0       20       511        0                   0   
WALKING                  0        0         1      483                   0   
WALKING_DOWNSTAIRS       0        0         0        1                 416   
WALKING_UPSTAIRS         0        0         1        2                   2   

Pred                WALKING_UPSTAIRS  
True                                  
LAYING                             0  
SITTING                            1  
STANDING                           1  
WALKING                           12  
WALKING_DOWNSTAIRS                 3  
WALKING_UPSTAIRS                 466  


In [56]:
from sklearn.metrics import accuracy_score

print('Train accuracy:',accuracy_score(np.argmax(Y_train, axis =1),predict(X_train)))
print('Test accuracy: ',accuracy_score(np.argmax(Y_test, axis = 1),predict(X_test)))

Train accuracy: 0.9880304678998912
Test accuracy:  0.9681031557516118


> By using two seperate models for static and dynamic activities, we got Test accuracy of 96.8%.

## <center> End of Self Case Study  